pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,346 @@
+    # not SO(4). Again we can fix this by simply negating a column.
+    q_in_so4 = qml.math.isclose(qml.math.linalg.det(q), 1.0)
+    if not q_in_so4:
+        q[:, -1] = -q[:, -1]
+
+    # Now, we should have p, q in SO(4) such that p^T u u^T p = q^T v v^T q.
+    # Then (v^\dag q p^T u)(v^\dag q p^T u)^T = I.
+    # So we can set G = p q^T, H = v^\dag q p^T u to obtain G v H = u.
+    G = qml.math.dot(p, qml.math.T(q))
+    H = qml.math.linalg.multi_dot([qml.math.conj(qml.math.T(v)), q, qml.math.T(p), u])
+
+    # These are still in SO(4) though - we want to convert things into SU(2) x SU(2)
+    # so use the entangler. Since u = E^\dagger U E and v = E^\dagger V E where U, V
+    # are the target matrices, we can reshuffle as in the docstring above,
+    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D)
+    # where A, B, C, D are in SU(2) x SU(2).
+    AB = qml.math.linalg.multi_dot([E, G, Edag])
+    CD = qml.math.linalg.multi_dot([E, H, Edag])
+
+    # Now, we just need to extract the constituent tensor products.
+    A, B = _su2su2_to_tensor_products(AB)
+    C, D = _su2su2_to_tensor_products(CD)
+
+    return A, B, C, D
+
+
+def two_qubit_decomposition(U, wires):
+    r"""Recover the decomposition of a two-qubit matrix :math:`U` in terms of
+    elementary operations.
+
+    The work of `Shende, Markov, and Bullock (2003)
+    <https://arxiv.org/abs/quant-ph/0308033>`__ presents a fixed-form
+    decomposition of :math:`U` in terms of single-qubit gates and
+    CNOTs. Multiple such decompositions are possible (by choosing two of
+    {``RX``, ``RY``, ``RZ``}). Here we choose the ``RY``, ``RZ`` case (fig. 2 in
+    the above) to match with the default decomposition of the single-qubit
+    ``Rot`` operations as ``RZ RY RZ``. The form of the decomposition is:
+
+    .. figure:: ../../_static/two_qubit_decomposition.svg
+        :align: center
+        :width: 100%
+        :target: javascript:void(0);
+
+    where :math:`A, B, C, D` are :math:`SU(2)` gates.
+
+    Args:
+        U (tensor): A 4 x 4 unitary matrix.
+        wires (Union[Wires, Sequence[int] or int]): The wires on which to apply the operation.
+
+    Returns:
+        list[qml.Operation]: A list of operations that represent the decomposition
+        of the matrix U.
+    """
+    # First, test if we have a tensor product of two single-qubit operations. If
+    # so, we don't actually need to do a decomposition. To test this, we can
+    # check if Edag U E is in SO(4) because of the isomorphism between SO(4) and
+    # SU(2) x SU(2).
+    test_so4 = qml.math.linalg.multi_dot([Edag, U, E])
+    if qml.math.isclose(qml.math.linalg.det(test_so4), 1.0) and qml.math.allclose(
+        qml.math.dot(test_so4, qml.math.T(test_so4)), qml.math.eye(4)
+    ):
+        A, B = _su2su2_to_tensor_products(U)
+        A_ops = zyz_decomposition(A, wires[0])
+        B_ops = zyz_decomposition(B, wires[1])
+        return A_ops + B_ops
```

Comment:
oh this is awesome!

Reply:
Actually I've been finding that the algorithm as detailed below does not work for some special cases, and that's why I've separated this one out. (It also fails if you try to put something like CNOT in). I think it has to do with the number of CNOTs required (it is at most 3, but may be less). I might leave this as a future TODO to improve the transform, though, since I figure most applications of `QubitUnitary` are for cases where we have some strange SU(4) op we know little enough about that we are using `QubitUnitary` in the first place.

Reply:
Oh, if you have a failing case, you could always add a test marked as `xfail` so that we have it for future dev purposes